cupy/cuda/memory.pyx

```diff
@@ -135,48 +135,33 @@ cdef class MemoryAsync(BaseMemory):
 
     Args:
         size (int): Size of the memory allocation in bytes.
-        stream (intptr_t): Pointer to the stream on which the memory is
-            allocated and freed.
+        stream (Stream): The stream on which the memory is allocated and freed.
     """
+
     cdef:
-        readonly intptr_t stream
+        readonly object stream_ref
 
-    def __init__(self, size_t size, intptr_t stream):
+    def __init__(self, size_t size, stream):
         self.size = size
         self.device_id = device.get_device_id()
         # The stream is allowed to be destroyed before the memory is freed, so
-        # we don't need to hold a reference to the stream.
-        self.stream = stream
-        is_async_alloc_supported(self.device_id)
+        # we don't need to hold a strong reference to the stream.
+        self.stream_ref = weakref.ref(stream)
+        self.ptr = 0
```

Comment:
Just a nitpick: cdef class attribute is zero-initialized, so we don't have to add this line.
```suggestion
```